### PR TITLE
WIP - Feature/grantfulltext

### PR DESCRIPTION
**Pull request by bahadasx**
Addresses #29 #30 #31 
Created new bash script to download and unzip grants and pubs files.  Also created new python script that handles parsing of Grants and Pubs data in addition to PTAB data.  This script replaces the existing script that handles PTAB parsing (parse_xml.py).  

I have not actually tried to send any Grants and Pubs files to Solr since the cores have not been set up yet, but the only difference should be the specific core name referenced in the request which my script is accounting for.

### CODE REVIEW

**Comment by bahadasx** [comment]
@arowla This PR is now ready to be reviewed.

File: parse_xml.py [hunk 1]
```diff
@@ -21,115 +22,207 @@ def changeExt(fname, ext):
...
+        try:
+            datetime.strptime(s, fmt)
+            return s
+        except ValueError:
+            pass
+    msg = "Not a valid date: '{0}'.".format(s)
+    raise argparse.ArgumentTypeError(msg)
+
+#replace key in dictionary
+def replaceKey(data,newval,oldval):
```
Comment (arowla) [inline parse_xml.py]: You don't need to fix this now, but stylistically, Python code should have commas followed by a space. The style with no spaces would set off an error in most style checkers (but [apparently it's not actually specified in PEP 8](http://stackoverflow.com/questions/31741418/pep8-space-after-a-comma)).

File: parse_xml.py [hunk 2]
```diff
@@ -21,115 +22,207 @@ def changeExt(fname, ext):
...
-                    x['APPLICANT_PUB_AUTHORIZATION_DT'] = formatDate(x, 'APPLICANT_PUB_AUTHORIZATION_DT')
-                    x['appid'] = x.pop('BD_PATENT_APPLICATION_NO')
-                    x['doc_date'] = x.pop('DOCUMENT_CREATE_DT')
-                    with open(txtfn) as dr:
-                        text = dr.read()
-                        x['textdata'] = text
+        fn = changeExt(fname,'json')
+        if not os.path.isfile(fn):
+            with open(fname) as fd:
+                logging.info("-- Beginning read of file at: "+str(datetime.now()))
```
Comment (arowla) [inline parse_xml.py]: Prefer to just embed the timestamp in the logging format. Requires a little bit of setup. From the [Python Logging Cookbook](https://docs.python.org/3/howto/logging-cookbook.html):

```
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
handler.setFormatter(formatter)
```

File: parse_xml.py [hunk 3]
```diff
@@ -21,115 +22,207 @@ def changeExt(fname, ext):
...
+                doc = xmltodict.parse(fd.read())
+                logging.info("-- End of reading file and converting to dictionary at: "+str(datetime.now()))
+                if args.ftype in ("g","p"):
```
Comment (arowla) [inline parse_xml.py]: Aim for more consistency in single vs double quotes. Generally use single unless there is string interpolation going on within the string (so the vast majority of strings should be single-quoted). See: http://stackoverflow.com/questions/56011/single-quotes-vs-double-quotes-in-python

File: parse_xml.py [hunk 4]
```diff
@@ -21,115 +22,207 @@ def changeExt(fname, ext):
...
+                docvalue = ["us-bibliographic-data-"+filetype[3],"publication-reference","document-id","appid"]
+                logfn = fname.split("_alt.json",1)[0]+"_solrlog.txt"
+            else:
+                element = "DATA_RECORD"
+                docvalue = ["DOCUMENT_IMAGE_ID"]
+                logfn = os.path.splitext(fname)[0]+"_solrlog.txt"
+            for x in doc['main'][element]:
+                x2 = x
+                for val in docvalue:
+                    x2 = x2[val]
```
Comment (arowla) [inline parse_xml.py:171]: How does this work? It seems like you could run into errors. `x2` is being reassigned a scalar value in each loop, yet you're assigning it _from_ `x2` on each go? Seems like only the first hit would work.

**Comment by arowla** [comment]
The grants retrieval file looks to be a copy-paste of the PTAB retrieval file. What are the chances the duplicated stuff could be factored out into another bash script and then `source`d?

**Pull request closed without merging**